src/ucp/rma/basic_rma.c

```diff
@@ -26,142 +30,242 @@
+
+    UCP_EP_RESOLVE_RKEY_RMA(ep, rkey, req->send.lane, uct_rkey, rma_config);
+    status = ucp_progress_put_inner(req, uct_rkey, rma_config);
+
+    return status;
+}
+
+static inline ucs_status_t
+ucp_rma_blocking(ucp_ep_h ep, const void *buffer, size_t length, uint64_t remote_addr, 
+                 ucp_rkey_h rkey, ucp_ep_rma_config_t *rma_config, 
+                 ucp_lane_index_t lane, uct_rkey_t uct_rkey,
+                 ucp_rma_send_func_t send_func)
+{
+    ucs_status_t status;
+    ucp_request_t req;
+
+    ucs_assert((send_func == ucp_progress_put_inner) ||
+               (send_func == ucp_progress_get_inner));
+
+    status = ucp_rma_request_init(&req, ep, buffer, length, remote_addr, rkey,
+                                  NULL, lane, 0);
+    if (status != UCS_OK) {
         return status;
     }
+
+    /* Loop until all message has been sent.
+     * We re-check the configuration on every iteration except for zcopy, 
+     * because it can be * changed by transport switch.
+     */
+    for (;;) {
+        status = send_func(&req, uct_rkey, rma_config);
+        if (ucs_likely(status == UCS_OK)) {
+            break;
+        } else if (status == UCS_INPROGRESS) {
+            continue;
+        } else if (status != UCS_ERR_NO_RESOURCE) {
+            break;
+        } else {
+            ucp_worker_progress(ep->worker);
+            UCP_EP_RESOLVE_RKEY_RMA(ep, rkey, req.send.lane, uct_rkey, rma_config);
+        }
+    }
+
+    ucp_rma_request_wait(&req);
+    return status;
 }
 
-static ucs_status_t
-ucp_rma_start_nbi(ucp_ep_h ep, const void *buffer, size_t length,
-                  uint64_t remote_addr, ucp_rkey_h rkey,
-                  uct_pending_callback_t cb)
+static inline ucs_status_t
+ucp_rma_nbi(ucp_ep_h ep, const void *buffer, size_t length, uint64_t remote_addr, 
+            ucp_rkey_h rkey, ucp_ep_rma_config_t *rma_config,
+            ucp_lane_index_t lane, uct_rkey_t uct_rkey,
+            ucp_rma_send_func_t send_func)
 {
+    ucs_status_t status;
+    ucs_status_t ret_status = UCS_INPROGRESS;
     ucp_request_t *req;
+    int can_send_more;
+
+    ucs_assert((send_func == ucp_progress_put_inner) ||
+               (send_func == ucp_progress_get_inner));
 
     req = ucp_request_get(ep->worker);
     if (req == NULL) {
         return UCS_ERR_NO_MEMORY;
     }
 
-    req->flags                = UCP_REQUEST_FLAG_RELEASED; /* Implicit release */
-    req->send.ep              = ep;
-    req->send.buffer          = buffer;
-    req->send.length          = length;
-    req->send.rma.remote_addr = remote_addr;
-    req->send.rma.rkey        = rkey;
-    req->send.uct.func        = cb;
-#if ENABLE_ASSERT
-    req->send.cb              = NULL;
-    req->send.lane            = UCP_NULL_LANE;
-#endif
-    return ucp_request_start_send(req);
+    ucp_rma_request_init(req, ep, buffer, length, remote_addr, rkey, 
+                         send_func == ucp_progress_put_inner ? 
```

Comment:
why not just pass ucp_progress_put_nbi / ucp_progress_get_nbi tho this function? it's ok if need to define new function type..